### PR TITLE
Fix bugs (core)

### PR DESCRIPTION
**Pull request by leventov**
Fix some bugs, most notably in `CostBalancerStrategy`. Other are performance/redundancy fixes, not logical bugs.

### CODE REVIEW

**Comment by drcrallen** [comment]
@leventov can you split these out into other PRs? they touch a lot of random pieces in code and seem to be missing related tests.

**Comment by leventov** [comment]
@drcrallen I don't see the point of splitting this into 5 different PRs, it increases noise for no benefit. Added a test to `GoogleDataSegmentPusher`. Only change in `RocketMQFirehoseFactory` is also theoretically testable, but I don't know how to use it.

**Comment by drcrallen** [comment]
Because its easier to merge quickly small PRs instead of having them all blocked on all of them passing code review.

**Comment by drcrallen** [comment]
And the commit history for extensions vs core becomes cleaner. So when the history of extensions is viewed, the changes for that extension are well isolated and clear.

I'm more interested in the logical separation of

1. Druid core
1. Extensions
1. When looking at the commit history affecting a file, does it make sense, or does it seem like it was changed in a bunch of random pull requests?
1. Anything that has API changes that can affect extensions.

I think 1, 2, and 3 come into play here.

File: processing/src/main/java/io/druid/query/ordering/StringComparators.java [hunk 1]
```diff
@@ -61,7 +62,7 @@ public int compare(String s, String s2)
     public int compare(String s, String s2)
     {
       // Avoid conversion to bytes for equal references
-      if(s == s2){
+      if(Objects.equals(s, s2)){
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/ordering/StringComparators.java]: Is this really better?

For strings that _are_ equal this is faster but for strings that aren't, it should be slower. I'd guess that most comparisons are between strings that are not equal, so the old code should perform better.

File: processing/src/main/java/io/druid/query/ordering/StringComparators.java [hunk 2]
```diff
@@ -374,7 +375,7 @@ private static BigDecimal convertStringToBigDecimal(String input) {
     public int compare(String o1, String o2)
     {
       // return if o1 and o2 are the same object
-      if (o1 == o2) {
+      if (Objects.equals(o1, o2)) {
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/ordering/StringComparators.java]: Similar question: is this really better?

File: extendedset/src/main/java/io/druid/extendedset/intset/AbstractIntSet.java [hunk 3]
```diff
@@ -114,7 +114,7 @@ public boolean containsAll(IntSet c)
     IntIterator itr = c.iterator();
     boolean res = true;
     while (res && itr.hasNext()) {
-      res &= contains(itr.next());
+      res = contains(itr.next());
```
Comment (gianm) [inline extendedset/src/main/java/io/druid/extendedset/intset/AbstractIntSet.java]: This wasn't a bug, right? Looks like behavior has stayed the same.

File: extendedset/src/main/java/io/druid/extendedset/intset/ConciseSet.java [hunk 4]
```diff
@@ -1324,7 +1323,7 @@ private boolean replaceWith(ConciseSet other)
     boolean isSimilar = (this.lastWordIndex == other.lastWordIndex)
                         && (this.last == other.last);
     for (int i = 0; isSimilar && (i <= lastWordIndex); i++) {
-      isSimilar &= this.words[i] == other.words[i];
+      isSimilar = this.words[i] == other.words[i];
```
Comment (gianm) [inline extendedset/src/main/java/io/druid/extendedset/intset/ConciseSet.java]: Same, this wasn't a bug, right?

Comment (leventov) [inline extendedset/src/main/java/io/druid/extendedset/intset/ConciseSet.java]: Not a bug

File: extendedset/src/main/java/io/druid/extendedset/intset/AbstractIntSet.java (hunk 3 above)
Comment (leventov) [inline extendedset/src/main/java/io/druid/extendedset/intset/AbstractIntSet.java]: Not a bug, redundancy

File: processing/src/main/java/io/druid/query/ordering/StringComparators.java (hunk 1 above)
Comment (leventov) [inline processing/src/main/java/io/druid/query/ordering/StringComparators.java]: Reverted

File: processing/src/main/java/io/druid/query/ordering/StringComparators.java (hunk 2 above)
Comment (leventov) [inline processing/src/main/java/io/druid/query/ordering/StringComparators.java]: Reverted

**Pull request merged**